bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/firmware/FirmwareUpdateServiceImpl.java

```diff
@@ -0,0 +1,556 @@
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.smarthome.config.core.validation.ConfigDescriptionValidator;
+import org.eclipse.smarthome.config.core.validation.ConfigValidationException;
+import org.eclipse.smarthome.core.common.SafeCaller;
+import org.eclipse.smarthome.core.common.ThreadPoolManager;
+import org.eclipse.smarthome.core.events.Event;
+import org.eclipse.smarthome.core.events.EventFilter;
+import org.eclipse.smarthome.core.events.EventPublisher;
+import org.eclipse.smarthome.core.events.EventSubscriber;
+import org.eclipse.smarthome.core.i18n.LocaleProvider;
+import org.eclipse.smarthome.core.i18n.TranslationProvider;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.firmware.Firmware;
+import org.eclipse.smarthome.core.thing.binding.firmware.FirmwareUpdateBackgroundTransferHandler;
+import org.eclipse.smarthome.core.thing.binding.firmware.FirmwareUpdateHandler;
+import org.eclipse.smarthome.core.thing.binding.firmware.ProgressCallback;
+import org.eclipse.smarthome.core.thing.events.ThingStatusInfoChangedEvent;
+import org.eclipse.smarthome.core.thing.firmware.FirmwareEventFactory;
+import org.eclipse.smarthome.core.thing.firmware.FirmwareRegistry;
+import org.eclipse.smarthome.core.thing.firmware.FirmwareStatus;
+import org.eclipse.smarthome.core.thing.firmware.FirmwareStatusInfo;
+import org.eclipse.smarthome.core.thing.firmware.FirmwareUpdateService;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Modified;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * Default implementation of {@link FirmwareUpdateService}.
+ *
+ * @author Thomas Höfer - Initial contribution
+ * @author Dimitar Ivanov - update and cancel operations are run with different safe caller identifiers in order to
+ *         execute asynchronously
+ * @author Dimitar Ivanov - Firmware update is done for thing
+ */
+@Component(immediate = true, service = { EventSubscriber.class, FirmwareUpdateService.class })
+public final class FirmwareUpdateServiceImpl implements FirmwareUpdateService, EventSubscriber {
+
+    private static final String THREAD_POOL_NAME = FirmwareUpdateServiceImpl.class.getSimpleName();
+    private static final Set<String> SUPPORTED_TIME_UNITS = ImmutableSet.of(TimeUnit.SECONDS.name(),
```

Comment:
`Collections.unmodifiableSet(Stream.of(...).toSet())` - see https://www.eclipse.org/smarthome/documentation/development/guidelines.html#c-language-levels-and-libraries

Reply:
Thanks for reminding me to use standard java 8 instead of guava.